Natural attack improvements

anatoliy-savchak opened the pull request:
* Made attack 4 account now;
* Attack 4 now accounts size modifier;
* Weapon Finesse workaround to skip str negative bonus;

DudeMcDude commented on TemplePlus/dispatcher.cpp:
```diff
@@ -19,6 +19,8 @@ class DispatcherReplacements : public TempleFix {
 public:
 
 	static int Dispatch54AoE(objHndl, DispIO *, D20DispatcherKey);
+	static int GetNumNaturalAttacks_(objHndl handle);
```
Why is this in dispatcher.cpp?? It has nothing to do with dispatchers...

Usually you can use the raw address to find a better place for it.

In this case it's obviously part of critterSys.

DudeMcDude commented on TemplePlus/dispatcher.cpp:
```diff
@@ -72,6 +77,40 @@ int DispatcherReplacements::Dispatch54AoE(objHndl handle, DispIO* evtObj, D20Dis
 
+int DispatcherReplacements::GetNumNaturalAttacks_(objHndl handle) {
+	if (!handle) return 0;
+	auto obj = objSystem->GetObject(handle);
```
For temple+ please also check obj.
Although rare, we sometimes get invalid non-null handles, which return a null object.

DudeMcDude commented on TemplePlus/dispatcher.cpp:
```diff
@@ -72,6 +77,40 @@ int DispatcherReplacements::Dispatch54AoE(objHndl handle, DispIO* evtObj, D20Dis
+	int idx = 0;
+	int attacksCount;
+	do
+	{
```
do{} while is bad form. It's an artifact of the decompiler, there's no need to adhere to it.

DudeMcDude commented on TemplePlus/dispatcher.cpp:
```diff
@@ -72,6 +77,40 @@ int DispatcherReplacements::Dispatch54AoE(objHndl handle, DispIO* evtObj, D20Dis
+	auto obj = objSystem->GetObject(handle);
+	int v2 = 0;
+	int result = 0;
+	while (1)
```
Likewise, decompiler produces crappy formatting. What seems like an infinite loop is really just a simple for loop.

int __cdecl GetDamageIdx(ObjHndl ObjHnd, int attackIdx){
int attackIdxSum = 0;
for (auto idx = 0; idx < 5; idx++){
    attackIdxSum += obj->GetInt32(obj_f_critter_attacks_idx, idx);
    if ( attackIdxSum > attackIdx )
      return idx;
  }
  return 0; // attackIdx is bigger than the sum of all obj_f_critter_attacks_idx items

anatoliy-savchak replied:
I wanted to change as little as possible. Thanks for advice. Will fix

DudeMcDude replied:
Also, it looks like I already rewrote this function!

And in critterSys too :)

DudeMcDude replied:
On that note. Since Temple+ already replaces a whole bunch of stuff, you should be extra careful when replacing existing functions. Sometimes doing a replaceFunction won't help because of exactly what happened here: I rewrote this function and called it explicitly in the code. That's my fault really, but that's how it is right now...

DudeMcDude commented on TemplePlus/critter.cpp:
```diff
@@ -1810,6 +1815,56 @@ void LegacyCritterSystem::BuildRadialMenu(objHndl handle){
 		}
 	}
 }
+int LegacyCritterSystem::SetCritterAttacks(objHndl handle)
```
This belongs in protos.cpp

anatoliy-savchak replied:
Funny, I first created in protos, but then tried to match the pattern, therefore moved here :)

Will fix, thanks

DudeMcDude commented on TemplePlus/condition.cpp:
```diff
@@ -476,6 +476,23 @@ class ConditionFunctionReplacement : public TempleFix {
 		redirectCall(0x100FC6D4, AnimalCompanionLevelHook);
+
+		// fix for negative str mod still accounting in the bon list
+		static int(*orgWeaponFinesseToHitBonus)(DispatcherCallbackArgs) =
```
One of my life's biggest regrets is this long-ass file :)
Please put it in ability_fixes.cpp

anatoliy-savchak commented:
Added new commit with such changes:
* Moved SetCritterAttacks from critter.cpp to protos.cpp;
* Moved GetNumNaturalAttacks from dispatcher.cpp to critter.cpp;
* Updated GetNumNaturalAttacks to use "for" loop;
* Deprecated GetDamageIdx hook as already implemented;
* Moved orgWeaponFinesseToHitBonus from condition.cpp to ability_fixes.cpp;

anatoliy-savchak commented:
No, its fine: from 0 to 3, therefore <4. See original:
  do
  {
    v3 = getArrayFieldInt32(obj, obj_f_critter_attacks_idx, v2);
    if ( v3 > 0 )
      v1 += v3;
    ++v2;
  }
  while ( v2 < 3 );

anatoliy-savchak commented:
![image](https://user-images.githubusercontent.com/40269424/61222128-bf329280-a722-11e9-8951-a6383e8f4ede.png)

DudeMcDude commented on TemplePlus/protos.cpp:
```diff
@@ -431,3 +435,54 @@ int ProtosHooks::ParseSpell(int colIdx, objHndl handle, char* content, obj_f fie
+		int32_t sizeCategory = obj->GetInt32(obj_f_size);
+		int sizeMod = critterSys.GetBonusFromSizeCategory(sizeCategory);
+		int attackIndex = 0;
+		do
```
Please fix this

Also please us obj->IsCritter() instead of explicit type==13 || 14

DudeMcDude commented on TemplePlus/protos.cpp:
```diff
@@ -431,3 +435,54 @@ int ProtosHooks::ParseSpell(int colIdx, objHndl handle, char* content, obj_f fie
+
+		auto dexAbility = obj->GetInt32(obj_f_critter_abilities_idx, 1);
+		result = objects.GetModFromStatLevel(dexAbility);
+		for (int i = result; attackIndex < 4; ++attackIndex)
```
This too

anatoliy-savchak replied:
Set Critter Attacks is handy function to remove str from damage dice and from attack bonus, because it will later be added during actual attack.

But the last 4th natural attach seems to be reserved specifically for Weapon Finesse attacks. It only removes dex mod from attack bonus. But it does not remove str mod from damage dice. Which is odd, as it still is going to affect damage dice.

**I suggest add str damage bonus removal similarly to 3 first natural attacks.** 

Arguments:
1) 4th attack is almost never used in protos. Exception - one dragon and one dude.
2) 4th attack even if it was specified in the protos was actually never used in vanilla (thus one of fixes above);
3)  Let's say I have Small Centipede with 1d4-3 damage. Without this fix I will need to remember that in 4 attack I need to manually change -3 to 0 (to simply 1d4), otherwise in reality during damage it will be 1d4-6 (as str mod is -3).

DudeMcDude commented:
Maybe it's meant for rangef attack?

anatoliy-savchak commented:
> Maybe it's meant for rangef attack?

I have not thought about that. OK, then I will not change it. Thanks!
